kubeadm: Make self-hosting work and split it out to a phase

luxas opened the pull request:
**What this PR does / why we need it**:

 - Removes the old self-hosting code
 - Puts the new self-hosting code in `phases/selfhosting`
   - The new code reads manifests from disk (static pods)...
   - ...mutates the PodSpec as necessary...
   - ...and posts the DaemonSet to the API Server...
   - ...and waits for it to come up
 - Uses DaemonSets for all control plane components
 - Creates a `kubeadm alpha phase selfhosting` command that can be invoked against any kubeadm-cluster after install.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

fixes: https://github.com/kubernetes/kubeadm/issues/127
(large part of at least)

**Special notes for your reviewer**:

Please only review the fourth commit, based on https://github.com/kubernetes/kubernetes/pull/47345

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
kubeadm: Make self-hosting work by using DaemonSets and split it out to a phase that can be invoked via the CLI
```
@kubernetes/sig-cluster-lifecycle-pr-reviews @jbeda

timothysc commented:
So this stuff needs to be merged in order, so marking this one as do-not-merge till the others are ready.

timothysc commented on cmd/kubeadm/app/cmd/phases/selfhosting.go:
```diff
@@ -0,0 +1,44 @@
+	kubeconfigutil "k8s.io/kubernetes/cmd/kubeadm/app/util/kubeconfig"
+)
+
+func NewCmdSelfhosting() *cobra.Command {
```
We should follow the conventions of // godoc on any public facing f(n)s.

timothysc commented on cmd/kubeadm/app/phases/selfhosting/podspec_mutation.go:
```diff
@@ -0,0 +1,103 @@
+	kubeadmconstants "k8s.io/kubernetes/cmd/kubeadm/app/constants"
+)
+
+func MutatePodSpec(name string, podSpec *v1.PodSpec) {
```
same.

timothysc commented on cmd/kubeadm/app/phases/selfhosting/podspec_mutation.go:
```diff
@@ -0,0 +1,103 @@
+	}
+}
+
+func addFileLockToPodSpec(podSpec *v1.PodSpec) {
```
I'm not entirely clear what you are doing here and why?

timothysc commented on cmd/kubeadm/app/phases/selfhosting/selfhosting.go:
```diff
@@ -0,0 +1,116 @@
+		// Wait for the self-hosted component to come up
+		kubeadmutil.WaitForPodsWithLabel(client, component, fmt.Sprintf("k8s-app=self-hosted-%s", component))
+
+		manifestPath := buildStaticManifestFilepath(component)
```
Could you comment this, it's totally unclear to anyone not working in this area what you are doing, and reference your phases.

timothysc commented on cmd/kubeadm/app/phases/selfhosting/selfhosting.go:
```diff
@@ -0,0 +1,116 @@
+
+		manifestPath := buildStaticManifestFilepath(component)
+		if err := os.RemoveAll(manifestPath); err != nil {
+			return fmt.Errorf("unable to delete static pod manifest for %s [%v]", component, err)
```
Can't we now potentially leave it in a 1/2 deployed state?  and future creates could fail?

timothysc commented on cmd/kubeadm/app/phases/selfhosting/selfhosting.go:
```diff
@@ -0,0 +1,116 @@
+}
+
+func buildDaemonSet(name string) (*extensions.DaemonSet, error) {
+	podSpec, err := loadPodSpecFromDisk(name)
```
Again comment on why you are loading from the original static manifest.

timothysc commented on cmd/kubeadm/app/phases/selfhosting/selfhosting.go:
```diff
@@ -0,0 +1,116 @@
+
+	return &extensions.DaemonSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "self-hosted-" + name,
```
Do we want to add either a label or name that it was kubeadm that did it?

luxas replied:
I don't know, what do you prefer?

timothysc commented on cmd/kubeadm/app/phases/selfhosting/selfhosting_test.go:
```diff
@@ -0,0 +1,18 @@
+*/
+
+// TODO: unit tests
+package selfhosting
```
Why even have this?  ¯\_(ツ)_/¯

luxas replied:
I'll add unit tests shortly

timothysc commented on cmd/kubeadm/app/util/apiclient.go:
```diff
@@ -52,3 +53,30 @@ func WaitForAPI(client *clientset.Clientset) {
+// WaitForPodsWithLabel will lookup pods with the given label and wait until they are all
+// reporting status as running.
+func WaitForPodsWithLabel(client *clientset.Clientset, name, labelKeyValPair string) {
+	wait.PollInfinite(kubeadmconstants.APICallRetryInterval, func() (bool, error) {
```
Shouldn't we poll timeout in case something weird happens on the kubelet?  B/c nothing weird ever happens on the kubelet.

luxas replied:
What would you prefer? How long should we wait before timing out?

kargakis replied:
How about a timeout flag in the selfhosted command?

luxas replied:
Can be a future improvement

kargakis replied:
Can you open an issue so we won't forget? Assign it to me if you like and I should get to it soon.

GheRivero commented on cmd/kubeadm/app/phases/selfhosting/podspec_mutation.go:
```diff
@@ -0,0 +1,103 @@
+package selfhosting
+
+import (
+	"k8s.io/client-go/pkg/api/v1"
```
Moved recently to k8s.io/api/core/v1

luxas commented:
- Rebased
 - Improved the code
 - Commented all functions
 - Added lots of unit tests (in their own commit for easier reviewing) 

In fact, the self-hosting code decreases with this PR with ~80 LoC.
This PR might look scary with a lot of changed lines, but in fact the actual code decreases.

However, the self-hosting code didn't have any comments or unit tests before, but now I've added ~700 LoC of unit tests :).
It's also possible to cleanup more code after this PRs is merged, but saving those cleanup tasks to a future PR

cc @jamiehannaford PTAL

kargakis commented on cmd/kubeadm/app/phases/selfhosting/podspec_mutation.go:
```diff
@@ -0,0 +1,74 @@
+	kubeadmconstants "k8s.io/kubernetes/cmd/kubeadm/app/constants"
+)
+
+// MutatePodSpec makes a Static Pod-hosted PodSpec suitable for self-hosting
```
nit: lowercase

kargakis commented on cmd/kubeadm/app/phases/selfhosting/selfhosting.go:
```diff
@@ -0,0 +1,143 @@
+// 8. In order to avoid race conditions, we're still making sure the API /healthz endpoint is healthy
+// 9. Do that for the kube-apiserver, kube-controller-manager and kube-scheduler in a loop
+func CreateSelfHostedControlPlane(client *clientset.Clientset) error {
+	components := []string{kubeAPIServer, kubeControllerManager, kubeScheduler}
```
Comment that the sequence matters or that it doesn't.

timothysc replied:
Do we want to treat them as separate Pods, or containers as part of the same pod?  

I'm not certain if we ever discussed a tradeoff here?  

The reason I ask is it simplifies the model.

jamiehannaford replied:
Maybe folks would want to scale more apiservers than schedulers/CMs, or vice versa? Just a thought

luxas replied:
I think it makes a lot of sense to decouple as three different DaemonSets (==3 different Pods) instead of having one DS and one Pod per master.

kargakis commented on cmd/kubeadm/app/phases/selfhosting/selfhosting.go:
```diff
@@ -0,0 +1,143 @@
+
+		// Create the DaemonSet in the API Server
+		if _, err := client.ExtensionsV1beta1().DaemonSets(metav1.NamespaceSystem).Create(ds); err != nil {
+			return fmt.Errorf("failed to create self-hosted %q daemonset [%v]", componentName, err)
```
Do we want to tolerate AlreadyExists errors?

kargakis commented on cmd/kubeadm/app/util/apiclient.go:
```diff
@@ -52,3 +54,29 @@ func WaitForAPI(client *clientset.Clientset) {
+		}
+		for _, pod := range apiPods.Items {
+			fmt.Printf("[apiclient] Pod %s status: %s\n", pod.Name, pod.Status.Phase)
+			if pod.Status.Phase != v1.PodRunning {
```
Do you prefer adding the /healthz check inside WaitForAPI as a readiness probe inside the static pod or by mutating its spec and then wait for Ready pods here instead of Running?

jamiehannaford replied:
We might need to disable healthz check on the bootstrap apiserver for self-hosted etcd due to [various race conditions](https://github.com/kubernetes-incubator/bootkube/issues/495), so I think this is the best solution for time being

jamiehannaford commented on cmd/kubeadm/app/phases/selfhosting/podspec_mutation.go:
```diff
@@ -0,0 +1,74 @@
+			setMasterTolerationOnPodSpec,
+			setRightDNSPolicyOnPodSpec,
+		},
+	}
```
For self-hosted etcd we'll also need to modify the podSpec `command` (mainly the apiserver). I assume we can pass in the additional flags which either overwrite or append to the existing one.

luxas replied:
Yes. I have coded some methods for modifying `.command` but didn't include that in this PR for easier reviewing

jamiehannaford commented on cmd/kubeadm/app/phases/selfhosting/podspec_mutation.go:
```diff
@@ -0,0 +1,74 @@
+}
+
+// addNodeSelectorToPodSpec makes Pod require to be scheduled on a node marked with the master label
+func addNodeSelectorToPodSpec(podSpec *v1.PodSpec) {
```
👍  Really like this composable way of mutating the original podSpec

jamiehannaford commented on cmd/kubeadm/app/phases/selfhosting/podspec_mutation.go:
```diff
@@ -0,0 +1,74 @@
+
+	podSpec.NodeSelector[kubeadmconstants.LabelNodeRoleMaster] = ""
+}
+
```
Do the self-hosted scheduler and controller manager really need to be on the host network? I think we should only use it when absolutely necessary because it uses up port space unnecessarily. If not, can we add a new mutator?

luxas replied:
> Do the self-hosted scheduler and controller manager really need to be on the host network

Yes they do, since there is no Pod Network available

> If not, can we add a new mutator?

I can add one to ensure `hostNetwork=true`

jamiehannaford replied:
Really? I thought pod IPs were addressable without a SDN. I got it working without flannel or anything. Maybe we can shelve this for now then

jamiehannaford commented on cmd/kubeadm/app/phases/selfhosting/selfhosting.go:
```diff
@@ -0,0 +1,143 @@
+// 1. Load the Static Pod specification from disk (from /etc/kubernetes/manifests)
+// 2. Extract the PodSpec from that Static Pod specification
+// 3. Mutate the PodSpec to be compatible with self-hosting (add the right labels, taints, etc. so it can schedule correctly)
+// 4. Build a new DaemonSet object for the self-hosted component in question. Use the PodSpec above mentioned PodSpec
```
s/Use the PodSpec above mentioned PodSpec/Use the above mentioned PodSpec/

jamiehannaford commented on cmd/kubeadm/app/phases/selfhosting/selfhosting.go:
```diff
@@ -0,0 +1,143 @@
+// 2. Extract the PodSpec from that Static Pod specification
+// 3. Mutate the PodSpec to be compatible with self-hosting (add the right labels, taints, etc. so it can schedule correctly)
+// 4. Build a new DaemonSet object for the self-hosted component in question. Use the PodSpec above mentioned PodSpec
+// 5. Create the DaemonSet resource. Wait for the Pod to become Running.
```
Wait for the Pods to be become Running

timothysc replied:
Wait until the Pods are running ;-)

jamiehannaford commented on cmd/kubeadm/app/phases/selfhosting/selfhosting.go:
```diff
@@ -0,0 +1,143 @@
+// 4. Build a new DaemonSet object for the self-hosted component in question. Use the PodSpec above mentioned PodSpec
+// 5. Create the DaemonSet resource. Wait for the Pod to become Running.
+// 6. Remove the Static Pod manifest file. The kubelet will stop the original Static Pod-hosted component that was running.
+// 7. The self-hosted container should now step up and take over.
```
containers

jamiehannaford commented on cmd/kubeadm/app/phases/selfhosting/selfhosting.go:
```diff
@@ -0,0 +1,143 @@
+	return nil
+}
+
+// buildDaemonSet is responsible for mutating the PodSpec and return a DaemonSet based suitable for the self-hosting purporse
```
which is suitable

jamiehannaford commented on cmd/kubeadm/app/phases/selfhosting/selfhosting.go:
```diff
@@ -0,0 +1,143 @@
+	// Return a DaemonSet based on that Spec
+	return &extensions.DaemonSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "self-hosted-" + name,
```
can we create a utility function to prepend the prefix? it's a bit brittle doing it manually all the time IMO. I see you added `buildSelfHostedWorkloadLabelQuery` already

timothysc replied:
+1 for mutating function with strings pushed to consts.

jamiehannaford commented on cmd/kubeadm/app/util/apiclient.go:
```diff
@@ -52,3 +54,29 @@ func WaitForAPI(client *clientset.Clientset) {
+			fmt.Printf("[apiclient] Error getting %s pods [%v]\n", name, err)
+			return false, nil
+		}
+		fmt.Printf("[apiclient] Found %d %s pods\n", len(apiPods.Items), name)
```
is it worth outputting when 0 pods are found? if not, can we move this beneath the `len(apiPods.Items) == 0` check

jamiehannaford replied:
We already log _a lot_ of info so being more picky would be nice 😄

jamiehannaford commented on cmd/kubeadm/app/phases/selfhosting/selfhosting_test.go:
```diff
@@ -0,0 +1,544 @@
+    - --kubeconfig=/etc/kubernetes/controller-manager.conf
+    - --controllers=*,bootstrapsigner,tokencleaner
+    - --root-ca-file=/etc/kubernetes/pki/ca.crt
+    - --address=127.0.0.1
```
can we change this arg to `0.0.0.0`. that will allow us to take off host network. same for scheduler

luxas replied:
Hmm, I'm not sure why that would allow us to remove `hostNetwork`...
As I understand this, the current state is "use the host's network, but don't expose anything widely"
We still need hostNetwork for being able to run things without CNI set up.

jamiehannaford commented on cmd/kubeadm/app/phases/selfhosting/selfhosting_test.go:
```diff
@@ -0,0 +1,544 @@
+    livenessProbe:
+      failureThreshold: 8
+      httpGet:
+        host: 127.0.0.1
```
also, can we make this an empty string? it defaults to pod IP if left blank. I see no reason to hardcode localhost here, especially if we want to move off host network

luxas replied:
I'd rather not do that in this PR. It's set in `app/phases/controlplane/manifests.go` since a long time back.

This is just testing data, but based on actual output from kubeadm.

luxas replied:
Also ref this statement: https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/

> Here’s one scenario where you would set it. Suppose the Container listens on 127.0.0.1 and the Pod’s hostNetwork field is true. Then host, under httpGet, should be set to 127.0.0.1.

To me it seems like it's required for hostNetwork=true

jamiehannaford replied:
Yeah, but my point is that the scheduler and controller manager don't need to be on host network. For apiserver (which is on host network), leaving blank will still work because it'll just ping that pod's API

luxas replied:
They must be on the host network due to that CNI is uninitialized, so we don't have any other option here really

jamiehannaford commented on cmd/kubeadm/app/phases/selfhosting/selfhosting_test.go:
```diff
@@ -0,0 +1,544 @@
+	}
+}
+
+func createTempFileWithContent(content []byte) (string, error) {
```
minor nit: I don't think unit tests should be modify external state - whether it be filesystem or DB. But due to the current state of our e2e tests maybe it's better to put here instead. What do you think?

luxas replied:
I've seen lots of unit tests that create temp files and then remove them.
A quick search for `ioutil.TempFile(` for files `kubernetes/*_test.go` yielded 170 results.
It's a pattern I definitely think we can use

jamiehannaford commented on cmd/kubeadm/app/phases/selfhosting/podspec_mutation.go:
```diff
@@ -0,0 +1,74 @@
+		kubeControllerManager: {
+			addNodeSelectorToPodSpec,
+			setMasterTolerationOnPodSpec,
+			setRightDNSPolicyOnPodSpec,
```
I noticed bootkube's [self-hosted controller manager](https://github.com/kubernetes-incubator/bootkube/blob/master/pkg/asset/internal/templates.go#L495) uses `Default` as the DNS policy. @aaronlevy is there any reason for that? Just want to make sure there's no corner cases we're forgetting about.

aaronlevy replied:
Looking back at the git blame I believe it had to do with the controller-manager needing to use the dns of the host when cloud-providers were enabled (for example, aws hostnames won't resolve unless you're using the host defined dns server). And at that time `DNSClusterFirstWithHostNet` didn't exist -- but seems like a more reasonable choice now (opened a PR to change in bootkube: https://github.com/kubernetes-incubator/bootkube/pull/629)

jamiehannaford commented on cmd/kubeadm/app/phases/selfhosting/selfhosting.go:
```diff
@@ -0,0 +1,143 @@
+		// Wait for the self-hosted component to come up
+		kubeadmutil.WaitForPodsWithLabel(client, componentName, buildSelfHostedWorkloadLabelQuery(componentName))
+
+		// Remove the old Static Pod manifest
```
Can we put all the teardown (L79-87) into a new function and have the init call it after CreateSelfHostedControlPlane? The reason I ask is that it'll need to happen after the etcd-operator deployment and pivot, so doing it now will reduce the future delta

luxas replied:
Discussed with @jamiehannaford on Slack and decided to hold off that for now, and re-order the sequence instead. This is fine as-is now and we'll change it if we change it in the future.

jamiehannaford reviewed (changes requested):
Looks great overall, just a few minor issues

luxas commented on cmd/kubeadm/app/phases/selfhosting/selfhosting.go:
```diff
@@ -0,0 +1,143 @@
+			Name:      "self-hosted-" + name,
+			Namespace: metav1.NamespaceSystem,
+			Labels: map[string]string{
+				"k8s-app": selfHostingLabelValuePrefix + name,
```
I don't think so, as I don't recognize any practical purpose for it
The less code the better

timothysc reviewed (changes requested):
Please address mine and other comments.  

poke again when done.

luxas commented on cmd/kubeadm/app/phases/selfhosting/podspec_mutation.go:
```diff
@@ -0,0 +1,74 @@
+// MutatePodSpec makes a Static Pod-hosted PodSpec suitable for self-hosting
+func mutatePodSpec(name string, podSpec *v1.PodSpec) {
+	mutators := map[string][]func(*v1.PodSpec){
+		kubeAPIServer: {
```
Hmm, I'm not sure what you mean... This works

luxas commented:
@timothysc @jamiehannaford @kargakis PTAL, updated the PR

luxas commented:
/retest

jamiehannaford reviewed (approved):
👍

kargakis commented on cmd/kubeadm/app/phases/selfhosting/selfhosting.go:
```diff
@@ -70,11 +73,17 @@ func CreateSelfHostedControlPlane(client *clientset.Clientset) error {
+				return fmt.Errorf("failed to create self-hosted %q daemonset [%v]", componentName, err)
+			}
+
+			if _, err := client.ExtensionsV1beta1().DaemonSets(metav1.NamespaceSystem).Update(ds); err != nil {
```
We may fail on status updates in the DaemonSet so I would make this retriable on 409s. TODO is also fine.

kargakis commented on cmd/kubeadm/app/phases/selfhosting/selfhosting.go:
```diff
@@ -70,11 +73,17 @@ func CreateSelfHostedControlPlane(client *clientset.Clientset) error {
+			}
+
+			if _, err := client.ExtensionsV1beta1().DaemonSets(metav1.NamespaceSystem).Update(ds); err != nil {
+				return fmt.Errorf("failed to create self-hosted %q daemonset [%v]", componentName, err)
```
failed to update

timothysc reviewed (approved):
/lgtm

luxas committed:
kubeadm: Make self-hosting work and split out to a phase

luxas committed:
kubeadm self-hosting: unit tests and bazel

luxas commented:
Squashed to two logical comments. Reapplying the labels

luxas commented:
/retest
